fix output data type of image classification

jiqing-feng opened the pull request:
Hi @Narsil . I found an issue when running the low-precision pipeline of image-classification:
```python
import torch
import requests
from transformers import pipeline
import PIL.Image

IMG_URL = "http://images.cocodataset.org/val2017/000000039769.jpg"
model_id = "google/vit-base-patch16-224"
classifier = pipeline("image-classification", model=model_id, torch_dtype=torch.bfloat16)
image = PIL.Image.open(requests.get(IMG_URL, stream=True, timeout=3000).raw)

output = classifier(image)
```

Error log:
```
Traceback (most recent call last):
  File "test_image_pipeline.py", line 11, in <module>
    output = classifier(image)
  File "/home/jiqingfe/miniconda3/envs/ipex/lib/python3.8/site-packages/transformers/pipelines/image_classification.py", line 157, in __call__
    return super().__call__(images, **kwargs)
  File "/home/jiqingfe/miniconda3/envs/ipex/lib/python3.8/site-packages/transformers/pipelines/base.py", line 1243, in __call__
    return self.run_single(inputs, preprocess_params, forward_params, postprocess_params)
  File "/home/jiqingfe/miniconda3/envs/ipex/lib/python3.8/site-packages/transformers/pipelines/base.py", line 1251, in run_single
    outputs = self.postprocess(model_outputs, **postprocess_params)
  File "/home/jiqingfe/miniconda3/envs/ipex/lib/python3.8/site-packages/transformers/pipelines/image_classification.py", line 183, in postprocess
    outputs = outputs.numpy()
TypeError: Got unsupported ScalarType BFloat16
```

We need to convert the output tensor to `float32` so it can be converted to `numpy`.

jiqing-feng commented:
Also, for the failed CI, it seems like not all tensors can be converted to `fp32`?

jiqing-feng committed:
fix output data type of image classification

amyeroberts commented on src/transformers/pipelines/image_classification.py:
```diff
@@ -179,7 +180,7 @@ def postprocess(self, model_outputs, function_to_apply=None, top_k=5):
             top_k = self.model.config.num_labels
 
-        outputs = model_outputs["logits"][0]
+        outputs = model_outputs["logits"][0].to(torch.float32)
         outputs = outputs.numpy()
```
We can't do this directly, as the outputs might not be torch tensors. 

We'll need to do something like: 
```suggestion
        outputs = model_outputs["logits"][0]
        
        if self.framework == "pt":
            # If torch_dtype is bf16 or fp16
            outputs = outputs.float().numpy()
        else:
            outputs = outputs.numpy()
```

amyeroberts reviewed (commented):
Thanks for working on a fix for this! 

We'll need to protect the float cast and add tests to check the pipeline works when `torch_dtype` is set to `float16` or `bfloat16`

jiqing-feng commented:
Hi @amyeroberts . Thanks for your review. I have fixed your comments and added a test for the low-precision pipeline, but the CI seems to be going wrong. Can you help check the test? Thx!

jiqing-feng committed:
add tests for low-precision pipeline

jiqing-feng commented:
Hi @amyeroberts. Can you take a look at the new changes? If it is okay for you, I will update all other pipelines to support low precision.

jiqing-feng commented:
Hi @amyeroberts . Could you please take a review? The failed CIs are not related to my changes :)

I am waiting for your comments to change the rest of the pipelines to support it. Thx!

jiqing-feng committed:
Merge branch 'huggingface:main' into image-classification

jiqing-feng commented:
Hi @amyeroberts . I have rebased and passed all CIs. Would you please review them?

amyeroberts commented on tests/pipelines/test_pipelines_image_classification.py:
```diff
@@ -177,6 +177,20 @@ def test_custom_tokenizer(self):
         self.assertIs(image_classifier.tokenizer, tokenizer)
 
+    @require_torch
+    def test_low_precision_pipeline(self):
```
Let's update to a more specific test name and then add an equivalent torch.bfloat16 

```suggestion
    def test_torch_float16_pipeline(self):
```

amyeroberts reviewed (approved):
Thanks for adding! 

Just a small request to add a test for bfloat16 too

jiqing-feng commented:
Hi @amyeroberts . I have added `fp16` and `bf16` tests for the image classification pipeline; could you please review them? Thx!

amyeroberts commented on tests/pipelines/test_pipelines_image_classification.py:
```diff
@@ -177,6 +177,34 @@ def test_custom_tokenizer(self):
 
+    @require_torch
+    def test_torch_float16_pipeline(self):
+        import torch
+
```
Instead of importing here - we should import at the top of the module with: 

```py
if is_torch_available():
    import torch
```

```suggestion
```

amyeroberts commented on tests/pipelines/test_pipelines_image_classification.py:
```diff
@@ -177,6 +177,34 @@ def test_custom_tokenizer(self):
+
+    @require_torch
+    def test_torch_bfloat16_pipeline(self):
+        import torch
+
```
```suggestion
```

amyeroberts commented on tests/pipelines/test_pipelines_image_classification.py:
```diff
@@ -177,6 +177,34 @@ def test_custom_tokenizer(self):
+        outputs = image_classifier("http://images.cocodataset.org/val2017/000000039769.jpg")
+
+        self.assertEqual(
+            nested_simplify(outputs, decimals=4),
```
I think?

```suggestion
            nested_simplify(outputs, decimals=3),
```

amyeroberts reviewed (approved):
Thanks for fixing! 

Just some small nits and we're ready to merge!

jiqing-feng committed:
Update tests/pipelines/test_pipelines_image_classification.py

Co-authored-by: amyeroberts <22614925+amyeroberts@users.noreply.github.com>

jiqing-feng committed:
add bf16 pipeline tests

jiqing-feng committed:
fix bf16 tests

jiqing-feng commented:
HI @amyeroberts . I have fixed the import issue, please take a review, thx!

jiqing-feng committed:
fix import

jiqing-feng committed:
fix import torch

jiqing-feng committed:
fix style
